src/bosh-director/lib/bosh/director/problem_resolver.rb

```diff
@@ -27,20 +35,20 @@ def track_and_log(task, log = true)
 
     def apply_resolutions(resolutions)
       @resolutions = resolutions
-      problems = Models::DeploymentProblem.where(id: resolutions.keys)
+      all_problems = Models::DeploymentProblem.where(id: resolutions.keys, deployment_id: @deployment.id).all
+      begin_stage('Applying problem resolutions', all_problems.size)
 
-      begin_stage('Applying problem resolutions', problems.count)
+      if Config.parallel_problem_resolution && all_problems.size > 1
+        ig_to_problems = problems_by_instance_group(all_problems)
 
-      if Config.parallel_problem_resolution
-        ThreadPool.new(max_threads: Config.max_threads).wrap do |pool|
-          problems.each do |problem|
-            pool.process do
-              process_problem(problem)
-            end
+        problems_serially_ordered_by_job(ig_to_problems) do |ig_problems, max_in_flight|
+          n_threads = [ig_problems.size, max_in_flight, Config.max_threads].min
+          parallel_each(n_threads, ig_problems) do |problem|
+            process_problem(problem)
```

Comment:
It feels confusing having`process_problem` being called in `process_ig`, which is being called by `problems_serially_ordered_by_job`. Cant this be refactored so that the ordering of the jobs, and their processing is done separately?
It seems that `apply_resolutions` should be the method calling the above methods.